src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -1690,20 +1690,47 @@ void JvmtiExport::continuation_yield_cleanup(JavaThread* thread, jint continuati
   }
 }
 
-void JvmtiExport::post_object_free(JvmtiEnv* env, jlong tag) {
-  Thread *thread = Thread::current();
+void JvmtiExport::post_object_free(JvmtiEnv* env, GrowableArray<jlong>* objects) {
+  assert(objects != NULL, "Nothing to post");
+  assert(env->is_enabled(JVMTI_EVENT_OBJECT_FREE), "checking");
+
+  Thread* thread = Thread::current();
+  if (thread->is_VM_thread()) {
+    post_object_free_on_vm_thread(env, objects);
+  } else {
+    post_object_free_on_java_thread(env, objects);
+  }
```

Comment:
The `post_object_free` can be called in different contexts.
It is (recursively) called from `JvmtiTagMap::check_hashmaps_for_heapwalk()` which is used in VM_ops:
- VM_HeapIterateOperation
- VM_HeapWalkOperation

Also, the `post_object_free` is called from `JvmtiTagMap` functions:
- flush_object_free_events
- get_objects_with_tags
which can be called from the `ServiceThread` or the JVMTI functions like `SetEvenNotificationMode`, `SetEventCallBacks` or `GetObjectsWithTags`.

Reply:
@sspitsyn  Thanks.

Yes, `JvmtiTagMap::check_hashmaps_for_heapwalk()` is where I saw the call from `VMThread`.

This patch is intended to **not** post `ObjectFree` events on `VMThread`, so I am going to move this PR back to draft state and take another look.

Reply:
> This patch is intended to not post ObjectFree events on VMThread, so I am going to move this PR back to draft state and take another look.

I'm not so sure it matters to the debug agent. I thought the main change here was to not force posting of ObjectFree events to be deferred to the VMThread, and to change the debug agent to send the CLASS_UNLOAD events right away rather than accumulating them and waiting for a GCFinishEvent. Is there any issue if the posting of ObjectFree and sending of CLASS_UNLOAD is done on the VMThread?